libraries/chain/controller.cpp

```diff
@@ -3958,34 +3955,32 @@ struct controller_impl {
       return fork_db.apply<std::optional<block_handle>>(unlinkable, f);
    }
 
-   // thread safe
+   // thread safe, QC already verified by verify_qc_claim
    void integrate_received_qc_to_block(const block_state_ptr& bsp_in) {
       // extract QC from block extension
+      assert(bsp_in->block);
       if (!bsp_in->block->contains_extension(quorum_certificate_extension::extension_id()))
          return;
 
       auto qc_ext = bsp_in->block->extract_extension<quorum_certificate_extension>();
-      const auto& received_qc = qc_ext.qc.data;
+      const qc_t& received_qc = qc_ext.qc;
 
-      const auto claimed = fork_db_fetch_bsp_on_branch_by_num( bsp_in->previous(), qc_ext.qc.block_num );
+      block_state_ptr claimed = fork_db_fetch_bsp_on_branch_by_num( bsp_in->previous(), qc_ext.qc.block_num );
       if( !claimed ) {
          dlog("qc not found in forkdb, qc: ${qc} for block ${bn} ${id}, previous ${p}",
               ("qc", qc_ext.qc.to_qc_claim())("bn", bsp_in->block_num())("id", bsp_in->id())("p", bsp_in->previous()));
          return;
       }
 
-      // Don't save the QC from block extension if the claimed block has a better or same valid_qc.
-      // claimed->valid_qc_is_strong() acquires a mutex.
-      if (received_qc.is_weak() || claimed->valid_qc_is_strong()) {
-         dlog("qc not better, claimed->valid: ${qbn} ${qid}, strong=${s}, received: ${rqc}, for block ${bn} ${id}",
-              ("qbn", claimed->block_num())("qid", claimed->id())("s", !received_qc.is_weak()) // use is_weak() to avoid mutex on valid_qc_is_strong()
+      // Don't save the QC from block extension if the claimed block has a better or same received_qc
+      if (!claimed->set_received_qc(received_qc)) {
+         dlog("qc not better, claimed->received: ${qbn} ${qid}, strong=${s}, received: ${rqc}, for block ${bn} ${id}",
+              ("qbn", claimed->block_num())("qid", claimed->id())("s", !received_qc.is_weak()) // use is_weak() to avoid mutex on received_qc_is_strong()
               ("rqc", qc_ext.qc.to_qc_claim())("bn", bsp_in->block_num())("id", bsp_in->id()));
          return;
```

Comment:
We should probably not `return` here as it makes sense to still do the `maybe_update_fsi()` below unless the incoming qc is exactlt the same we already had.

Reply:
Excellent point. That is a really good improvement over what we had before.